### DISCUSSION TITLE
Sub-select with ORM object changes in ORM-touching sub-select doesn't cause flush?

### DISCUSSION THREAD

**Discussion by igorham000** [sqlalchemy/sqlalchemy#9804] [Usage Questions]
I was debugging a hybrid_attribute on an orm'd class, and was getting strange results. The query was basically a simple select of a specific row of one table, with the only column being a sub-select grabbing a corresponding element on a different table - phrased as a sub-select to deal with a group-by clause. Something like

```sql
-- b.a_val will always collapse to 1 row when grouped by b.d.
-- a.val will be repeated many times for a given b, with different a.id's
SELECT
  (SELECT array_agg(a.id) FROM a WHERE a.val = b.a_val) as aggregation
FROM b GROUP BY b.d 
```

After poking around a bit, I found out it was because the ORM session wasn't flushing before query execution. I found this line seemed to be the key (orm/sessions.py:2027)
![image](https://github.com/sqlalchemy/sqlalchemy/assets/103453290/4f3eb7bf-596d-4574-83b7-27d1228ea254)

which, in something like `SELECT a.id FROM a`, causes the if branch on sessons.py:2097 to return True, causing the session to flush.
![image](https://github.com/sqlalchemy/sqlalchemy/assets/103453290/fb3cdf23-895a-4c3d-857b-198f5f77036f)

In the unhappy path - e.g. queries like above - the compile_state_cls is None because the "orm" plugin wasn't found in that first if statement, even though it's set in the sub-select.

To reproduce what I'm seeing:

```python
from sqlalchemy import literal, select

from sqlalchemy import MetaData, create_engine, engine, orm
from sqlalchemy.orm import DeclarativeBase, Mapped, Session, mapped_column


class Base(DeclarativeBase):
  metadata = MetaData()

class Foo(Base):
  __tablename__ = 'test'

  id: Mapped[int] = mapped_column(primary_key=True)
  val: Mapped[int]


def run():
  engine = create_engine('<YOUR TEST DB DSN HERE>')
  Base.metadata.create_all(engine)
  with Session(engine) as session:
   # Flush a Foo with (id=1, val=2) to the DB.
    f = Foo(id=1, val=2)
    session.add(f)
    # Update f.val to 3.
    f.val = 3

   # Each of the following use a that *should* hint that the ORM is being used
   # to the execute call, which will always return the above-created row.
   #
   # The first two examples stick that select in a subselect, and filter on f.val = 3, which
   # will only be true if the above update was flushed to the DB.

    # Situation 1: explicit flush before selecting subselect, 1 value returned.
    # session.flush()
    # print(session.scalars(select(
    #   select(Foo.id).where(Foo.id.is_not(literal(None))).scalar_subquery()
    # ).where(Foo.val == 3)).first())

    # Situation 2: No explicit flush before selecting subselect, no value returned.
    # print(session.scalars(select(
    #   select(Foo.id).where(Foo.id.is_not(literal(None))).scalar_subquery()
    # ).where(Foo.val == 3)).first())

    # Situation 3: No explicit flush before selecting w/o subselect, 1 value returned.
    # print(session.scalars(select(Foo.id)
    #       .where(Foo.id.is_not(literal(None)))
    #       .where(Foo.val == 3)).first())
if __name__ == '__main__':
  run()
```

Not really clear on the rules around flushing, or if this is expected behavior - the code suggested they were more nuanced than "flush before every query execution", but that's what the docs I found suggested to be the case 🤔

---

Tested on postgres against versions 2.0.13 & 2.0.14

**Comment by zzzeek** [comment]
I've added #9804 for this which will use ORM contexts for all statements that have any ORM attributes anywhere within them.  This might be too heavy-handed of a change depending on how the tests/ review go.    Ultimately, "autoflush" would be turned on for all statements within session.execute(), but this would be too big a change for 2.0.x.
